### ISSUE TITLE
Embedded discriminator subdoc properties cannot be set directly from parent doc

### ISSUE THREAD

**Issue by asabhaney** [Automattic/mongoose#6482]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

When trying to set a property on a embedded discriminator subdoc, there seems to be 2 bugs preventing the change from taking effect.

**If the current behavior is a bug, please provide the steps to reproduce.**

```js
const assert = require('chai').assert
const mongoose = require('mongoose')
const Schema = mongoose.Schema

it('Embedded discriminator subdoc properties cannot be set using `parentDoc.set`', async () => {

   // Setup our schemas + model
   // Everything works okay here

   const mediaSchema = new Schema(
      { file: String },
      { discriminatorKey: 'kind', _id: false }
   )

   const photoSchema = new Schema({ 
      position: String
   })

   const pageSchema = new Schema({
      media: mediaSchema
   })

   pageSchema.path('media').discriminator('photo', photoSchema)

   const Page = mongoose.model('test-page', pageSchema)

   // Create new Page doc
   // Everything cool here too...

   let page = new Page({
      media: {
         kind: 'photo',
         file: 'cover.jpg',
         position: 'left'
      }
   })

   await page.save()
   assert.equal(page.media.kind, 'photo')
   assert.equal(page.media.file, 'cover.jpg')
   assert.equal(page.media.position, 'left')

   // Try changing the page.media props
   // This is where things start to go off the rails

   // ** THIS FAILS **

   page.set('media.position', 'right')
   assert.equal(page.media.position, 'right')

   // ** BUT IT WORKS IF WE DO IT THIS WAY **
   // (Comment out the above assertion)

   page.media.set('position', 'right')
   assert.equal(page.media.position, 'right')

   // ** OH NO! ANOTHER BUG - IT DOESN'T SAVE BECAUSE NO MODIFIED PATHS DETECTED

   assert.equal(page.modifiedPaths().length > 0, true)
   assert.equal(page.modifiedPaths().includes('media.position'), true)

   // ** BUT IT WORKS IF WE EXPLICITLY MARK AS MODIFIED **
   // (Comment out the above 2 assertions)

   page.markModified('media.position')

   // Phew...we have our expected result!

   await page.save()
   page = await Page.findById(page.id)
   assert.equal(page.media.kind, 'photo')
   assert.equal(page.media.file, 'cover.jpg')
   assert.equal(page.media.position, 'right')
})
```

**Please mention your node.js, mongoose and MongoDB version.**

node.js 8.9.4
mongoose 5.1.1
MongoDB 3.6.4

**Comment by lineus** [comment]
Hey guys, I'm on a mission to learn as much as I can about discriminators, so I took to trying to sort out this issue, but I'm a little perplexed by the example. 

The `media` path in the `pageSchema` isn't an array. From my *albeit brief* experience and some time spent with the discriminator docs, it seems like it should be an array:
```
   const pageSchema = new Schema({
      media: mediaSchema
   })

   pageSchema.path('media').discriminator('photo', photoSchema)
```

**Comment by asabhaney** [comment]
@lineus support for discriminators in single embedded docs was added as a feature in 97397ccdbf2b0d6cf4332385b7cd8a2862adf0e7.

You bring up a good point though, this should probably be documented in the [Discriminators](http://mongoosejs.com/docs/discriminators.html) section of the docs. I can try to submit a PR that adds docs for this when I get some time.

**Comment by lineus** [comment]
Thanks @asabhaney!

**Comment by asabhaney** [comment]
Thanks for the fix @vkarpov15. Sorry, I should've tested for doc arrays as well.

I was trying out the latest master, and noticed that the above test case is now passing, but it fails again if I change the setter to use an object as the first param instead:

```js
page.set({ 'media.position': 'right' })
```

Here is the entire test case if it helps:

```js
const assert = require('chai').assert
const mongoose = require('mongoose')
const Schema = mongoose.Schema

it('Embedded discriminator subdoc properties cannot be set using `parentDoc.set`', async () => {

   // Setup our schemas + model
   // Everything works okay here

   const mediaSchema = new Schema(
      { file: String },
      { discriminatorKey: 'kind', _id: false }
   )

   const photoSchema = new Schema({
      position: String
   })

   const pageSchema = new Schema({
      media: mediaSchema
   })

   pageSchema.path('media').discriminator('photo', photoSchema)

   const Page = mongoose.model('test-page', pageSchema)

   // Create new Page doc
   // Works fine

   let page = new Page({
      media: {
         kind: 'photo',
         file: 'cover.jpg',
         position: 'left'
      }
   })

   await page.save()
   assert.equal(page.media.kind, 'photo')
   assert.equal(page.media.file, 'cover.jpg')
   assert.equal(page.media.position, 'left')

   // Start changing the page.media props

   // ** THIS FAILS **

   page.set({ 'media.position': 'right' })
   assert.equal(page.media.position, 'right')

   // ** BUT IT WORKS IF WE DO IT THIS WAY **
   // (Comment out the above assertion)

   page.set('media.position', 'right')
   assert.equal(page.media.position, 'right')
})
```

**Comment by vkarpov15** [comment]
Thanks, fixed :+1:

**Comment by vkarpov15** [comment]
All 3 fixes should be in 5.1.3, so please upgrade and let us know if you find any similar issues :+1:

**Comment by asabhaney** [comment]
Thanks @vkarpov15, seems to be working as expected now :)
